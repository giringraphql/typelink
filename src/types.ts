export type ResolvedValue<T> = T | Promise<T>;
export type ResolvedList<T> = T[] | Promise<T[]> | Promise<T>[];

export interface ConcreteClass<T = any> {
  new(...args: any[]): T;
}

export type Instantiator<TClass = any> = (value: { [key: string]: any }) => TClass;

export function defaultInputFieldInstantiator(value: any) {
  return value;
}

export type Thunk<T> = () => T;
