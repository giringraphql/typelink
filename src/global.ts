import { loadFallbackRootTypes } from './definition/rootTypes';
import { loadBuiltInScalar } from './definition/scalar';
import { DefinitionKind } from './metadata/Definition';
import { MetadataStorage } from './metadata/MetadataStorage';
import { DefinitionParser } from './sdl/ast';
import { coerceType } from './type-expression/coerceType';
import { TypeExpression } from './type-expression/TypeExpression';
import { Thunk } from './types';
import { TypeArg } from './type-expression/types';


/**
 * Create a new MetadataStorage initialized with default metadata
 */
export function createMetadataStorage() {
  const storage = new MetadataStorage();
  loadBuiltInScalar(storage);
  loadFallbackRootTypes(storage);
  return storage;
}

/**
 * Global MetadataStorage used by default.
 */
export let globalMetadataStorage: MetadataStorage;

export function getGlobalMetadataStorage() {
  if (!globalMetadataStorage) {
    globalMetadataStorage = createMetadataStorage();
  }
  return globalMetadataStorage;
}

/**
 * Get a GraphQLType instance from the given storage or default
 * global metadata storage.
 * @param typeArg
 * @param storage
 */
export function getType(typeArg: TypeExpression | TypeArg, kind: DefinitionKind = 'any'): any {
  const storage = getGlobalMetadataStorage();
  return coerceType(typeArg).getType({ kind, storage });
}

/**
 * Define a type linked to decorated class and add it to the given
 * storage or default global metadata storage.
 * @param metadataOrThunk
 */
export function defineType(parsersOrThunk: DefinitionParser[] | Thunk<DefinitionParser[]>) {
  const storage = getGlobalMetadataStorage();
  return function defDecoratorFn<T extends Function>(definitionClass: T) {
    if (Array.isArray(parsersOrThunk)) {
      parsersOrThunk.forEach(parser => {
        storage.registerMetadata(parser.parse(definitionClass));
      });
    } else {
      storage.deferRegister(() => {
        parsersOrThunk().forEach(parser => {
          storage.registerMetadata(parser.parse(definitionClass));
        });
      });
    }
    return definitionClass;
  };
}
