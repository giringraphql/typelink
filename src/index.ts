export { gql } from './sdl/gql';

export * from './definition/GraphQLTypeIndex';
export * from './definition/InputType';
export * from './definition/InterfaceType';
export * from './definition/ObjectType';
export * from './definition/rootTypes';
export * from './definition/scalar';
export * from './definition/SubscriptionType';


export * from './reference/Field';
export * from './reference/Implement';
export * from './reference/InputField';
export * from './global';

export * from './type-expression/DefinitionTypeExpression';
export * from './type-expression/GraphQLTypeExpression';
export * from './type-expression/TypeExpression';
export * from './type-expression/structure';
export * from './type-expression/coerceType';
export * from './type-expression/types';

export * from './metadata/Definition';
export * from './metadata/MetadataStorage';
export * from './metadata/Reference';

export * from './relay/connection';
export * from './relay/mutation';
export * from './relay/node';
