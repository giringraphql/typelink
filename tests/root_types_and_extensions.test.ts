import { GraphQLSchema, printSchema, graphql } from 'graphql';
import { defineType, gql, Mutation, Query, getType } from '../src';


@defineType(gql`
  type User {
    id: Int!
    username: String!
  }
  input UserInput {
    username: String!
  }

  extend type Query {
    getUser(userId: Int!): User!
    userClassName: String!
  }
  extend type Mutation {
    createUser(user: UserInput!): User!
  }
`)
class User {
  id: number;
  username: string;

  // extend type Query
  static getUser(_source: null, { userId }: { userId: number }) {
    const user = new User();
    user.id = userId;
    user.username = `User${userId}`;
    return user;
  }

  static userClassName() {
    return this.name;
  }

  // extend type Mutation
  static createUser(_source: null, { user }: { user: User }) {
    return user;
  }
}

const schema = new GraphQLSchema({
  query: getType(Query),
  mutation: getType(Mutation),
  types: [getType(User)],
});


describe('Root types and extensions', () => {
  it('should generate schema as expected', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('has resolvers binded to their definition class', async () => {
    const res = await graphql({ schema, source: `
      query {
        getUser(userId: 12) {
          id
          username
        }
        userClassName
      }
    `});
    expect(res).toEqual({
      data: {
        getUser: {
          id: 12,
          username: 'User12',
        },
        userClassName: 'User',
      },
    });
  });
});
