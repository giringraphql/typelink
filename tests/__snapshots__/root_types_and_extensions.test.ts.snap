// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root types and extensions should generate schema as expected 1`] = `
"type Mutation {
  createUser(user: UserInput!): User!
}

type Query {
  getUser(userId: Int!): User!
  userClassName: String!
}

type User {
  id: Int!
  username: String!
}

input UserInput {
  username: String!
}
"
`;
